src/Model/Scope/AbstractCondition.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+declare(strict_types=1);
+
+namespace atk4\data\Model\Scope;
+
+use atk4\core\InitializerTrait;
+use atk4\core\TrackableTrait;
+use atk4\data\Exception;
+use atk4\data\Model;
+
+/**
+ * @property CompoundCondition $owner
+ */
+abstract class AbstractCondition
+{
+    use InitializerTrait {
+        init as _init;
+    }
+    use TrackableTrait;
+
+    /**
+     * Method is executed when the scope is added to parent scope using Scope::add
+     * $this->owner is the Scope object.
+     */
+    public function init(): void
+    {
+        if (!$this->owner instanceof self) {
+            throw new Exception('Scope can only be added as element to scope');
+        }
+
+        $this->_init();
+
+        $this->onChangeModel();
+    }
+
+    abstract public function onChangeModel();
+
+    /**
+     * Temporarily assign a model to the condition.
+     *
+     * @return static
+     */
+    final public function on(Model $model)
+    {
+        $clone = clone $this;
+
+        $clone->owner = null;
+
+        (clone $model)->scope()->add($clone);
+
+        return $clone;
+    }
+
+    /**
+     * Get the model this condition is associated with.
+     */
+    public function getModel(): ?Model
+    {
+        return $this->owner ? $this->owner->getModel() : null;
+    }
+
+    /**
+     * Empty the scope object.
+     *
+     * @return static
+     */
+    abstract public function clear();
+
+    /**
+     * Negate the scope object
+     * e.g from 'is' to 'is not'.
+     *
+     * @return static
+     */
+    abstract public function negate();
+
+    /**
+     * Return if scope has any conditions.
+     */
+    abstract public function isEmpty(): bool;
+
+    /**
+     * Convert the scope to human readable words when applied on $model.
+     *
+     * @return bool
+     */
+    abstract public function toWords(bool $asHtml = false): string;
```

Comment:
I think `$asHtml` should be dropped

Reply:
Any other idea how to render the scope as string nicely so that field names are highlighted for instance?

Reply:
traverse over the Scope objects in ui and merge the result there

Reply:
Maybe define a renderer Object which takes ($field, $operator, $value) words and returns a string?

Reply:
yes - but not renderer, but helper method that will return these for your renderer, should should be responsible to traverse and render like renderer wants